### DISCUSSION TITLE
[Q] Websocket Server to Expose tracing events to other clients (like a javascript client for example)

### DISCUSSION THREAD

**Discussion by shekohex** [tokio-rs/tracing#1958] [Q&A] [answered]
Hello! I have a Real-Time System that mostly listens to events from the outside world sources and does some actions depending on these events. The system is implemented in Rust and we are using `tracing` and `tracing-subscriber` with `fmt` logging.

Our Integration tests system is implemented in Typescript (using jest and nodejs, due to this system is interacting or listing to these sources which are implemented in javascript.) and we want a way to inspect inside the state of our application by emitting some tracing events using the macros we already using from the `tracing` crate and push them to a WebSocket server for example! and assert on them on our integration tests (kinda similar to the concept of [Trace based testing](https://github.com/aspecto-io/malabi#tbt))

So, **Is there any crates that implement a tracing subscriber and expose a WebSocket server for that for example?**

Thank you very much for tracing crate!

**Comment by hawkw** [comment, marked as answer]
I don't believe anything like this currently exists, but it's certainly possible to implement. I think you could probably write a simple implementation of this by implementing the [`MakeWriter` trait](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/trait.MakeWriter.html) for a type that writes to a WebSocket connection. This would allow you to forward events logged by the `fmt` subscriber over WebSockets.

You may want to use the [JSON formatter](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/format/struct.Json.html) to format those events, so your TypeScript program can make assertions about their structure without having to parse a textual log line.

**Reply by shekohex** [reply]
> You may want to use the [JSON formatter](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/format/struct.Json.html) to format those events, so your TypeScript program can make assertions about their structure without having to parse a textual log line.

Yes! I found this the easy way to go, the typescript test is already starting the inner Rust executable and manage it through the lifetime of the testing, so reading the output in JSON will do it.

For anyone looking into doing something similar, without changing too much:
```rust
// ~snip~
// if we are not compiling for integration tests, we should use pretty logs
    #[cfg(not(feature = "integration-tests"))]
    let logger = logger.pretty();
    // otherwise, we should use json, which is easy to parse.
    #[cfg(feature = "integration-tests")]
    let logger = logger.json().flatten_event(true).with_current_span(false);

    logger.init();
```

and in Typescript Side:
```ts
// ~snip~
// use the `JSONStream` package, to parse the JSON stream for us.
this.#process.stdout
      ?.pipe(JSONStream.parse())
      .on('data', (parsedLog: UnparsedRawEvent) => {
        if (parsedLog.target === 'webb_probe') {
          const rawEvent = {
            timestamp: new Date(parsedLog.timestamp),
            ...(parsedLog as any),
          } as RawEvent;
          this.#logs.push(rawEvent);
          this.#eventEmitter.emit(rawEvent.kind, rawEvent);
        }
      });
```

**Reply by hawkw** [reply]
Cool! I'm glad this is working for you!
